### DISCUSSION TITLE
How to get Avalonia.Controls.Screens instance without access to an instance of WindowBase?

### DISCUSSION THREAD

**Discussion by fitdev** [AvaloniaUI/Avalonia#14067] [Q&A] [answered]
Hello!

I need to query available displays (screens) and their properties _before_ any UI such as `TopLevel`s or `Window`s are instantiated, so I cannot use `WindowBase.Screens` for that. So how do I do it?

I assume that `WindowBase.Screens` reports _all screens available on a system_, and NOT _only the screens_ where a part of the window is drawn (typically this would be just 1 screen). Correct me please if my assumption is wrong.

**Comment by thevortexcloud** [comment, marked as answer]
I don't think you can do this before Avalonia is ready? Someone else may correct me though. You can certainly do it by side stepping Avalonia and querying all the display information yourself/using another library. SDL2 for example should be able to provide all the information you need and is cross platform, but it's a C++ library so it can be not very intuitive to use for a C# dev. But there is [an excellent C# wrapper for it.](https://github.com/flibitijibibo/SDL2-CS)

Alternatively, can't you just wait until Avalonia is ready and start your windows invisible?

**Reply by fitdev** [reply]
Thank you for your reply and suggestion. Going for a 3rd party library is certainly an overkill for my needs. However, I think there should be a way to get certain Platform-specific information in Avalonia, such as screen information, without having to wait for some `TopLevel`-derived class to be instantiated, since that information may be need for various architectural needs before a window is actually instantiated. So if currently there is no such a way, perhaps there could be some API like `static IPlatformInfo PlatformInformation.GetCurrent()` or some such thing.

**Reply by timunie** [reply]
Nope as not supported by all platforms. Technically each TopLevel can have different screen setups. 

When exactly do you need that info? Probably you can access it during Window.Loaded event?

**Reply by fitdev** [reply]
Ideally before any of the UI logic runs. I just thought it is still rather strange to expose such Platform-level information via an instance of the UI like a window. The number of and capabilities of the desktop screens for example does not change whether you have 0, 1, or 5 applications windows open. And for mobile platforms, you might still be interested not in the number of screens, but in the capabilities of the main screen for example, and too need that info before any window is instantiated. So, I still think that it would be valuable if Avalonia could somehow statically expose that platform information without the need to create a window for that.

**Reply by fitdev** [reply]
Oh, and does `WindowBase.Screens` report _only the screens_ where a part of the window is drawn (typically this would be just 1 screen). Or does it report _all screens available on a system_? I am interested in the latter.

**Reply by thevortexcloud** [reply]
It reports the _active_ screen information of all screens. [It's loosely based on the Winforms version of the class.](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.screen?view=windowsdesktop-8.0)

Which is to say, it wont give you all mode information. If you need that, then you will need to get the information yourself or use a library like SDL2. The winforms implementation though is a static class, but they only have to worry about a single platform.

**Reply by timunie** [reply]
All screens the TopLevel can see. It's by design. So follow it or create your own logic. We are open source, so you can check the implementation

**Reply by thevortexcloud** [reply]
> and too need that info before any window is instantiated.

There is nothing stopping you from waiting until Avalonia has finished it's initialisation process and then instantiating all your windows. You don't have to create a window during boot, you can do it at some point after. It's just not a very common thing to do. But people do it for things like starting their app as a tray icon only, and then opening the window with the tray icon. But I am still confused why you need it before the window has initialised rather than during/after the window initialisation?

**Reply by fitdev** [reply]
Thank you for your suggestions! I guess I will just wait until main window is instantiated to query that information. Yeah, I was hoping for WinForms-like functionality.
